api/helpers.go

```diff
@@ -91,3 +108,196 @@ func paginationToRequest[T any](rows []T, dbPageSize int32, cursor string, goFor
+	if err != nil {
+		return nil, "", nil, err
+	}
+	// encode the holders
+	holdersAddresses, holdersValues := [][]byte{}, [][]byte{}
+	for addr, balance := range opts.Holders {
+		key := addr.Bytes()[:censustree.DefaultMaxKeyLen]
+		if opts.Type != anonymousCensusType {
+			if key, err = ref.Tree().Hash(addr.Bytes()); err != nil {
+				return nil, "", nil, err
+			}
+		}
+		holdersAddresses = append(holdersAddresses, key[:censustree.DefaultMaxKeyLen])
+		value := ref.Tree().BigIntToBytes(balance)
+		holdersValues = append(holdersValues, value)
+	}
+	// add the holders to the census tree
+	db.Lock()
+	if _, err := ref.Tree().AddBatch(holdersAddresses, holdersValues); err != nil {
+		return nil, "", nil, err
+	}
+	root, err := ref.Tree().Root()
+	if err != nil {
+		return nil, "", nil, err
+	}
+	data, err := ref.Tree().Dump()
+	if err != nil {
+		return nil, "", nil, err
+	}
+	db.Unlock()
+	// generate the tree dump
+	dump, err := censusdb.BuildExportDump(root, data, opts.Type, censustree.DefaultMaxLevels)
+	if err != nil {
+		return nil, "", nil, err
+	}
+	// publish it on IPFS
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	uri, err := storage.Publish(ctx, dump)
+	if err != nil {
+		return nil, "", nil, err
+	}
+	db.Lock()
+	if err := db.Del(bID); err != nil {
+		return nil, "", nil, err
+	}
+	db.Unlock()
+	return root, uri, dump, nil
+}
+
+// InnerCensusID generates a unique identifier by concatenating the BlockNumber, StrategyID,
+// and a numerical representation of the Anonymous flag from a CreateCensusRequest struct.
+// The BlockNumber and StrategyID are concatenated as they are, and the Anonymous flag is
+// represented as 1 for true and 0 for false. This concatenated string is then converted
+// to a uint64 to create a unique identifier.
+func InnerCensusID(blockNumber, strategyID uint64, anonymous bool) uint64 {
+	// Convert the boolean to a uint64: 1 for true, 0 for false
+	var anonymousUint uint64
+	if anonymous {
+		anonymousUint = 1
+	}
+	// Concatenate the three values as strings
+	concatenated := fmt.Sprintf("%d%d%d", blockNumber, strategyID, anonymousUint)
+	// Convert the concatenated string back to a uint64
+	result, err := strconv.ParseUint(concatenated, 10, 64)
+	if err != nil {
+		panic(err)
+	}
+	if result > math.MaxInt64 {
+		panic(err)
+	}
+	return result
+}
+
+// CalculateStrategyHolders function returns the holders of a strategy and the
+// total weight of the census. It also returns the total block number of the
+// census, which is the sum of the strategy block number or the last block
+// number of every token chain id. To calculate the census holders, it uses the
+// supplied predicate to filter the token holders using a lexer and evaluator.
+// The evaluator uses the strategy operators to evaluate the predicate which
+// uses the database queries to get the token holders and their balances, and
+// combines them.
```

Comment:
the block number is defined in the strategy_token table and will be used as part of result of `TokensByStrategyID` query method, i will do because right now the census generation for a certain block number is not supported.

Reply:
Its not, but it will be supported, right? So IMHO it would make sense to allow the caller specify a block number on this function. Anyway we can do it later on.